frappe/integrations/doctype/ldap_settings/test_ldap_settings.py

```diff
@@ -1,8 +1,1071 @@
 # -*- coding: utf-8 -*-
 # Copyright (c) 2019, Frappe Technologies and Contributors
 # See license.txt
-# import frappe
-import unittest
+import frappe
+import unittest, functools, ldap3, ssl, os
 
-class TestLDAPSettings(unittest.TestCase):
-	pass
+from unittest import mock
+from frappe.integrations.doctype.ldap_settings.ldap_settings import LDAPSettings
+from ldap3 import Server, Connection, MOCK_SYNC, OFFLINE_SLAPD_2_4, OFFLINE_AD_2012_R2, ALL_ATTRIBUTES
+
+
+class LDAP_TestCase():
+	TEST_LDAP_SERVER = None # must match the 'LDAP Settings' field option
+	TEST_LDAP_SEARCH_STRING = None
+	LDAP_USERNAME_FIELD = None
+	DOCUMENT_GROUP_MAPPINGS = []
+	LDAP_SCHEMA = None
+	LDAP_LDIF_JSON = None
+	TEST_VALUES_LDAP_COMPLEX_SEARCH_STRING = None
+
+	def mock_ldap_connection(f):
+
+		@functools.wraps(f)
+		def wrapped(self, *args, **kwargs):
+
+			#with mock.patch('ldap3.Connection') as mock_connection:
+			with mock.patch('frappe.integrations.doctype.ldap_settings.ldap_settings.LDAPSettings.connect_to_ldap') as mock_connection:
+				mock_connection.return_value = self.connection
+
+				self.test_class = LDAPSettings(self.doc)
+
+				# Create a clean doc
+				localdoc = self.doc.copy()
+				frappe.get_doc(localdoc).save()
+
+				rv = f(self, *args, **kwargs)
+
+
+			# Clean-up
+			self.test_class = None
+
+			return rv
+
+		return wrapped
+
+	def clean_test_users():
+		try: # clean up test user 1
+			frappe.get_doc("User", 'posix.user1@unit.testing').delete()
+		except:
+			pass
+
+		try: # clean up test user 2
+			frappe.get_doc("User", 'posix.user2@unit.testing').delete()
+		except:
+			pass
+
+
+	@classmethod
+	def setUpClass(self, ldapServer='OpenLDAP'):
+
+		print(self.TEST_LDAP_SERVER + ' setup')
+		self.clean_test_users()
+		# Save user data for restoration in tearDownClass()
+		#ToDo: bug. group mappings can be added even if they already exist. see if a row can be made unique
```

Comment:
Discovery:

Can the `ldap_group_mappings` document be made unique?

Currently, duplicate rows that match exactly can be created.

Reply:
implemented `ToDo` in commit 6592cdb1703aaf459746ccf1dce433d057e53fc3